rust/primary/src/messages.rs

```diff
@@ -1,189 +1,255 @@
-use super::committee::*;
-use super::error::*;
-use super::types::*;
-use crypto::{Digest, PublicKey, SecretKey, Signature};
-use ed25519_dalek::Digest as DalekDigest;
+use crate::error::{DagError, DagResult};
+use crate::primary::Round;
+use config::{Committee, WorkerId};
+use crypto::{Digest, Hash, PublicKey, Signature, SignatureService};
+use ed25519_dalek::Digest as _;
 use ed25519_dalek::Sha512;
 use serde::{Deserialize, Serialize};
-use std::collections::BTreeSet;
-use tokio::sync::mpsc::Sender;
-use tokio::sync::oneshot;
-
-#[cfg(test)]
-#[path = "tests/messages_tests.rs"]
-pub mod messages_tests;
-
-/// This is the data structure the receive re-computes.
-#[derive(Debug, Clone, Serialize, Deserialize)]
-pub struct BlockHeader {
-    /// The author of the block.
-    pub author: NodeID,
-    /// The round number of the block.
-    pub round: RoundNumber,
-    /// The digests of all transactions in that block.
-    pub transactions_digest: WorkersDigests,
-    /// Links to the parents of the block.
-    pub parents: BTreeSet<(NodeID, Digest)>,
-    /// User-defined metadata.
-    pub metadata: Metadata,
-    /// The current instance of the protocol.
-    pub instance_id: SequenceNumber,
+use std::collections::{BTreeMap, BTreeSet, HashSet};
+use std::convert::TryInto;
+use std::fmt;
+
+#[derive(Clone, Serialize, Deserialize, Default)]
+pub struct Header {
+    pub author: PublicKey,
+    pub round: Round,
+    pub payload: BTreeMap<Digest, WorkerId>,
+    pub parents: BTreeSet<Digest>,
+    pub id: Digest,
+    pub signature: Signature,
 }
-pub const GC_DEPTH: u64 = 10;
-
-impl BlockHeader {
-    /// Verify that the block header is well formed.
-    pub fn check(&self, committee: &Committee) -> Result<(), DagError> {
-        dag_ensure!(
-            self.instance_id == committee.instance_id,
-            DagError::BadInstanceId
-        );
-        Ok(())
-    }
 
-    // Checks if a block has some digest as parent
-    pub fn has_parent(&self, digest: &Digest) -> bool {
-        for (_, parent) in &self.parents {
-            if *parent == *digest {
-                return true;
-            }
+impl Header {
```

Comment:
We used to also keep a serialized version, do you think it was not helpful/faster?

Reply:
It didn't really make a difference, it was only making the header more complex to handle. Probably a better strategy is to keep in memory a serialized version of the header in the `Helper` to help other nodes to quickly sync